File: lib/features/scan/data/scan_service.dart
// ignore_for_file: prefer_initializing_formals
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:image/image.dart' as img;

import '../../../core/network/dictionary_service.dart';
import '../../../core/vision/marker_detector.dart';
import '../../../core/vision/ocr_service.dart';
import '../../word/data/word_generation_service.dart';
import '../../word/data/word_repository.dart';

final markerDetectorProvider = Provider((ref) => MarkerDetector());

final ocrServiceProvider = Provider<OcrService>((ref) {
  final service = OcrService();
  ref.onDispose(service.dispose);
  return service;
});

final scanServiceProvider = Provider<ScanService>((ref) {
  return ScanService(
    detector: ref.watch(markerDetectorProvider),
    ocr: ref.watch(ocrServiceProvider),
    dictionary: ref.watch(dictionaryServiceProvider),
    wordRepository: ref.watch(wordRepositoryProvider),
  );
});

class ScanCandidate {
  const ScanCandidate({
    required this.text,
    required this.needsReview,
    required this.isDuplicate,
  });

  /// OCRで認識された単語/フレーズ
  final String text;

  /// 辞書に見つからず読み取り不完全の疑いがある場合true
  final bool needsReview;

  /// 既存単語帳に同じ単語が存在する場合true
  final bool isDuplicate;
}

class ScanService {
  ScanService({
    required MarkerDetector detector,
    required OcrService ocr,
    required DictionaryService dictionary,
    required WordRepository wordRepository,
  })  : _detector = detector,
        _ocr = ocr,
        _dictionary = dictionary,
        _wordRepository = wordRepository;

  final MarkerDetector _detector;
  final OcrService _ocr;
  final DictionaryService _dictionary;
  final WordRepository _wordRepository;

  Future<List<ScanCandidate>> scan(img.Image image) async {
    final regions = _detector.detect(image);

    final texts = <String>[];
    for (final region in regions) {
      final cropped = _cropWithMinSize(image, region);
      if (cropped == null) continue;
      final text = await _ocr.recognizeText(cropped);
      if (text.isNotEmpty) texts.add(text);
    }

    // 候補内重複の集約（正規化した完全一致・大小文字無視）
    final seenNormalized = <String>{};
    final uniqueTexts = <String>[];
    for (final text in texts) {
      final trimmed = text.trim();
      final normalized = trimmed.toLowerCase();
      if (trimmed.isEmpty || seenNormalized.contains(normalized)) continue;
      seenNormalized.add(normalized);
      uniqueTexts.add(trimmed);
    }

    final candidates = <ScanCandidate>[];
    for (final text in uniqueTexts) {
      var needsReview = false;
      try {
        await _dictionary.lookup(text);
      } on WordNotFoundException {
        needsReview = true;
      }
      final existing = await _wordRepository.findByWord(text);
      candidates.add(
        ScanCandidate(
          text: text,
          needsReview: needsReview,
          isDuplicate: existing != null,
        ),
      );
    }
    return candidates;
  }
}

/// ML Kitは入力画像の幅・高さが32px未満だとエラーになるため、
/// 余白を付けつつ最小サイズを確保してクロップする
img.Image? _cropWithMinSize(img.Image image, MarkerRegion region) {
  const padding = 10;
  const minSize = 48;

  var left = region.left - padding;
  var top = region.top - padding;
  var width = region.width + padding * 2;
  var height = region.height + padding * 2;

  if (width < minSize) {
    left -= (minSize - width) ~/ 2;
    width = minSize;
  }
  if (height < minSize) {
    top -= (minSize - height) ~/ 2;
    height = minSize;
  }

  left = left.clamp(0, image.width - 1);
  top = top.clamp(0, image.height - 1);
  width = width.clamp(1, image.width - left);
  height = height.clamp(1, image.height - top);

  if (width < minSize || height < minSize) return null;

  return img.copyCrop(image, x: left, y: top, width: width, height: height);
}

File: web/src/services/scanService.ts
import { detectMarkerRegions, type MarkerRegion } from './markerDetector';
import { recognizeText } from './ocrService';
import { lookupWord, WordNotFoundError } from './dictionaryService';
import { findByWord } from '../db/wordRepository';

export interface ScanCandidate {
  /** OCRで認識された単語/フレーズ */
  text: string;
  /** 辞書に見つからず読み取り不完全の疑いがある場合true */
  needsReview: boolean;
  /** 既存単語帳に同じ単語が存在する場合true */
  isDuplicate: boolean;
}

/** ML Kit/Tesseractは極端に小さい画像だと精度が落ちるため、余白を付けつつ最小サイズを確保してクロップする */
const CROP_PADDING = 10;
const MIN_CROP_SIZE = 48;

function cropWithMinSize(source: HTMLCanvasElement, region: MarkerRegion): HTMLCanvasElement | null {
  let left = region.left - CROP_PADDING;
  let top = region.top - CROP_PADDING;
  let width = region.width + CROP_PADDING * 2;
  let height = region.height + CROP_PADDING * 2;

  if (width < MIN_CROP_SIZE) {
    left -= Math.floor((MIN_CROP_SIZE - width) / 2);
    width = MIN_CROP_SIZE;
  }
  if (height < MIN_CROP_SIZE) {
    top -= Math.floor((MIN_CROP_SIZE - height) / 2);
    height = MIN_CROP_SIZE;
  }

  left = Math.min(Math.max(left, 0), source.width - 1);
  top = Math.min(Math.max(top, 0), source.height - 1);
  width = Math.min(Math.max(width, 1), source.width - left);
  height = Math.min(Math.max(height, 1), source.height - top);

  if (width < MIN_CROP_SIZE || height < MIN_CROP_SIZE) return null;

  const cropped = document.createElement('canvas');
  cropped.width = width;
  cropped.height = height;
  const ctx = cropped.getContext('2d')!;
  ctx.drawImage(source, left, top, width, height, 0, 0, width, height);
  return cropped;
}

export type ScanProgress = { stage: 'detecting' | 'ocr' | 'lookup'; current: number; total: number };

/** 1枚の画像からマーカー領域を検出し、OCR・重複集約・辞書照合まで行う */
export async function scanImage(
  image: HTMLCanvasElement,
  onProgress?: (progress: ScanProgress) => void,
): Promise<ScanCandidate[]> {
  const regions = detectMarkerRegions(image);

  const texts: string[] = [];
  for (let i = 0; i < regions.length; i++) {
    onProgress?.({ stage: 'ocr', current: i + 1, total: regions.length });
    const cropped = cropWithMinSize(image, regions[i]);
    if (!cropped) continue;
    const text = await recognizeText(cropped);
    if (text) texts.push(text);
  }

  // 候補内重複の集約（正規化した完全一致・大小文字無視）
  const seenNormalized = new Set<string>();
  const uniqueTexts: string[] = [];
  for (const text of texts) {
    const trimmed = text.trim().replace(/\s+/g, ' ');
    const normalized = trimmed.toLowerCase();
    if (!trimmed || seenNormalized.has(normalized)) continue;
    seenNormalized.add(normalized);
    uniqueTexts.push(trimmed);
  }

  const candidates: ScanCandidate[] = [];
  for (let i = 0; i < uniqueTexts.length; i++) {
    onProgress?.({ stage: 'lookup', current: i + 1, total: uniqueTexts.length });
    const text = uniqueTexts[i];
    let needsReview = false;
    try {
      await lookupWord(text);
    } catch (e) {
      if (e instanceof WordNotFoundError) {
        needsReview = true;
      } else {
        // ネットワークエラー等、辞書照合そのものができない場合も要確認扱いにする
        needsReview = true;
      }
    }
    const existing = await findByWord(text);
    candidates.push({ text, needsReview, isDuplicate: !!existing });
  }
  return candidates;
}

/** 複数画像（PDFの各ページなど）をまとめてスキャンし、画像をまたいだ重複も除去する */
export async function scanImages(
  images: HTMLCanvasElement[],
  onProgress?: (imageIndex: number, imageTotal: number, progress: ScanProgress) => void,
): Promise<ScanCandidate[]> {
  const all: ScanCandidate[] = [];
  const seen = new Set<string>();
  for (let i = 0; i < images.length; i++) {
    const candidates = await scanImage(images[i], (p) => onProgress?.(i + 1, images.length, p));
    for (const c of candidates) {
      const key = c.text.toLowerCase();
      if (seen.has(key)) continue;
      seen.add(key);
      all.push(c);
    }
  }
  return all;
}
